# TCSS 478/588 In-class Exercise: Correlation and linear regression

These exercises will be used to illustrate how to perform data analyses using R.
Alternatively, you can use R or Rstudio. 
I assume you will learn basic R commands on your own.

Credit: The data and ideas behind these exercises and homeworks are from the NIH LINCS DCIC Crowdsourcing Portal and Ma’ayan Lab @ Mt Sinai, New York.
http://www.maayanlab.net/crowdsourcing/megatask1.php

The overarching goal is to predict adverse drug reactions. But we will take this task step by step. You will learn more about the meanings of these data in future lectures. For illustrative purposes in today’s class, just treat these data as numbers in a matrix.

## 1. Download and read in data (same as last week)
Download the data “gene_expression_n438x978.txt” from canvas to your computer.  I have removed the first character (#) and an ‘ from the input file to make R happy.

In [1]:
getwd()

[1] "C:/Users/radeshpa/Desktop/BioInformatics/Assignment3"

In [2]:
my.data <- read.table ("gene_expression_n438x978.txt", sep="\t", header=T)
dimnames(my.data)[[1]] <- my.data[,1]
my.data <- my.data[, -1]
my.data[1:5, 1:6]

PSME1        ATF1         RHEB         FOXO3      
ACETAZOLAMIDE       -0.015159099 -0.031470528 -0.004733488  0.02591061
IRBESARTAN          -0.026811981  0.012151979 -0.025550148 -0.02401181
IPRATROPIUM BROMIDE  0.001017958 -0.008650622 -0.018128698 -0.02079971
EFAVIRENZ           -0.004398264  0.055387992  0.004658520  0.01380732
THIAMINE             0.001838965 -0.018079188 -0.011855532 -0.03705033
                    RHOA          IL1B        
ACETAZOLAMIDE        0.0056296773 -0.007325439
IRBESARTAN          -0.0106717396  0.070951154
IPRATROPIUM BROMIDE -0.0002722781  0.025003360
EFAVIRENZ           -0.0340697348  0.024005885
THIAMINE            -0.0133954959 -0.001752450

In [3]:
dim (my.data)

[1] 438 978

In [8]:
res <- t(my.data)
dim(res)


[1] 978 438

## 2. Correlation 

In [5]:
help (cor)

In [11]:
correlation <- cor(my.data, use = "everything",
    method = c("pearson"))
correlation


PSME1         ATF1         RHEB          FOXO3         RHOA        
PSME1     1.000000000   0.026699663  0.033151809  -0.0126188574  0.001813023
ATF1      0.026699663   1.000000000  0.127226603  -0.1147250388  0.015006002
RHEB      0.033151809   0.127226603  1.000000000  -0.0486450461  0.119331159
FOXO3    -0.012618857  -0.114725039 -0.048645046   1.0000000000  0.088304142
RHOA      0.001813023   0.015006002  0.119331159   0.0883041419  1.000000000
IL1B     -0.009108597  -0.048410774 -0.179658776  -0.0767136128 -0.061326215
ASAH1    -0.096497245   0.199088611  0.112075730   0.0038381796  0.012853942
RALA     -0.012409473   0.139894125 -0.101306260  -0.1035809440 -0.114714261
ARHGEF12 -0.070925659  -0.064300902  0.029579953  -0.0863475938  0.151936855
SOX2     -0.084706615  -0.103227257 -0.063637341  -0.0556200975  0.034189136
SERPINE1 -0.046855628   0.001324413  0.038816161   0.0517187324 -0.025756263
HLA.DMA  -0.120916400  -0.096855685  0.049948620   0.0879424137  0.106252755
EGF      -0.012225738  -0.056897660  0.010903129  -0.0344866120 -0.066517819
APP      -0.004136425   0.072823848  0.066806653   0.0406934668 -0.075371604
NOS3     -0.097641269  -0.100700796 -0.005113953  -0.0016491297 -0.016161121
CSNK1A1  -0.015157547   0.046306417  0.112585292   0.1729851255  0.120722942
NFATC4   -0.188677073  -0.145530409 -0.216865677  -0.0497404195 -0.104424219
TBP       0.144313538   0.086453907  0.180716021  -0.0394624985  0.230474073
BRCA1    -0.004109635   0.128306217 -0.031451259  -0.1057762201 -0.139581116
PSMD4    -0.052660763  -0.071321285  0.068972920   0.0919611863  0.063728851
ETV1     -0.047409449   0.019648970 -0.061367185   0.0076237158  0.029535180
TERT     -0.115113973  -0.176913212 -0.059650869   0.0244466575 -0.012067568
EED       0.021870570   0.222615622 -0.060221584  -0.1512907638 -0.065748553
PTK2B    -0.154568091  -0.103779364 -0.094596202  -0.0392968712 -0.009973579
HSPB1    -0.042768695  -0.111000682  0.006417008   0.1653509780  0.012047157
PIK3C3   -0.005880439   0.061510733 -0.055269288   0.0148787057 -0.125286811
CBLB     -0.110123623  -0.009481546 -0.030009866   0.0173645378  0.081425367
DFFB     -0.005251343   0.023098458 -0.007162509  -0.0009275593  0.014144204
TGFB3    -0.149988181  -0.135667972 -0.093704037  -0.0719410459  0.015826163
PRKX     -0.067830410  -0.050654917 -0.139737539   0.0074975511 -0.044065324
...      ...           ...          ...           ...           ...         
KLHL9    -0.0654497296  0.018953831  0.1516068027 -0.035327306   0.090859524
UBE2J1   -0.0855841503 -0.156537777 -0.0598789027  0.032298059  -0.036976631
SLC37A4   0.1227208319 -0.022025682 -0.0003446413 -0.009959304   0.006270311
NPRL2     0.0116398271 -0.083323493 -0.0465747073 -0.019312562   0.024263478
PWP1     -0.0470632701 -0.068103625  0.0063926635  0.004013542   0.087980451
PLEKHM1  -0.0047278696 -0.066545661 -0.0959666392  0.049139970  -0.116344395
VAT1      0.0785244986 -0.037989732  0.0150176767  0.041020959  -0.035100467
KIF20A    0.0152660294  0.045486815  0.1596098074 -0.038296798   0.147290426
CDH3     -0.0022342553 -0.064894289 -0.0061040891  0.011038792  -0.069802561
SSBP2    -0.0904016242  0.005237222 -0.0596462235 -0.049311904   0.043502746
SNX11    -0.0106758812  0.001392690  0.0376923792  0.082295888   0.108736370
SFN      -0.1380508677 -0.087106333  0.0388347251 -0.020320593   0.098994495
RFX5      0.1353906168  0.015047651 -0.0222076321  0.147215948  -0.076851982
PKIG      0.0430838848  0.061698005 -0.0803004718 -0.013555781   0.025402724
FSD1     -0.0610166937 -0.107024983 -0.1017861151 -0.039174137  -0.037271055
ATP6V1D  -0.0153969538 -0.075786116 -0.0311669318  0.106483343   0.106253338
CDC45    -0.0480722364 -0.017374800  0.0444589404 -0.025508326   0.049790807
EDEM1    -0.0150094485 -0.055000926 -0.0052818173  0.109118947   0.022786119
USP6NL   -0.0534853450  0.048858409  0.0136881527 -0.004294719   0.005140586
NT5DC2   -0.0008391351 -0.096940175  0.0647930163  0.001716752   0.078607959
CCP11

## 3. Pairwis correlation

_Your task 1:_ What is the maximum correlation coefficient among all possible pairs of genes?  How about the minimum correlation?  

_Q:_ Can you get away from doing a double nested loop in R?
Discuss with your neighbors.

Note that the total number of possible gene pairs = 978*977/2 = 477,753

_Your task 2:_ write a R script. For computationally intensive tasks, you might want to consider running your R code in batch mode.  https://www.r-bloggers.com/batch-processing-vs-interactive-sessions/

Hint: There are many ways to do this. You can use a loop or you might want to check out
http://www.r-bloggers.com/using-apply-sapply-lapply-in-r/

In [12]:
max(correlation, na.rm = "TRUE")

[1] 1

## 4. Linear Regression.

In [14]:
X <- c(1, 0, -1, 0)
typeof(X)

[1] "double"

In [16]:
X <- c(1:10)

In [17]:
typeof(X)

[1] "integer"

In [23]:
X[c(T,T,F,F,F,F,F,F,F,F,F)]

[1] 1 2

In [24]:
install.packages("gapminder")

Installing package into 'C:/Users/radeshpa/Documents/R/win-library/3.5'
(as 'lib' is unspecified)
also installing the dependencies 'assertthat', 'utf8', 'cli', 'fansi', 'pillar', 'pkgconfig', 'rlang', 'tibble'



package 'assertthat' successfully unpacked and MD5 sums checked
package 'utf8' successfully unpacked and MD5 sums checked
package 'cli' successfully unpacked and MD5 sums checked
package 'fansi' successfully unpacked and MD5 sums checked
package 'pillar' successfully unpacked and MD5 sums checked
package 'pkgconfig' successfully unpacked and MD5 sums checked
package 'rlang' successfully unpacked and MD5 sums checked
package 'tibble' successfully unpacked and MD5 sums checked
package 'gapminder' successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\radeshpa\AppData\Local\Temp\RtmpCKgpLB\downloaded_packages


In [25]:
library(gapminder)

In [27]:
my.data(gapminder)

ERROR: Error in my.data(gapminder): could not find function "my.data"
